mobilecoind-json/src/main.rs

```diff
@@ -0,0 +1,402 @@
+    if let Some(value) = extra.value {
+        req.set_value(value);
+    }
+    if let Some(memo) = extra.memo.clone() {
+        req.set_memo(memo);
+    }
+
+    let resp = state
+        .mobilecoind_api_client
+        .get_request_code(&req)
+        .map_err(|err| format!("Failed getting request code: {}", err))?;
+
+    Ok(Json(JsonRequestCodeResponse {
+        request_code: String::from(resp.get_b58_code()),
+    }))
+}
+
+#[derive(Serialize, Default)]
+struct JsonReadRequestResponse {
+    value: String,
+    memo: String,
+}
+
+/// Retrieves the metadata in a request code
+#[get("/read-request/<request_code>")]
+fn read_request(
+    state: rocket::State<State>,
+    request_code: String,
+) -> Result<Json<JsonReadRequestResponse>, String> {
+    let mut req = mc_mobilecoind_api::ReadRequestCodeRequest::new();
+    req.set_b58_code(request_code.clone());
+    let resp = state
+        .mobilecoind_api_client
+        .read_request_code(&req)
+        .map_err(|err| format!("Failed reading request code: {}", err))?;
+
+    Ok(Json(JsonReadRequestResponse {
+        value: resp.get_value().to_string(),
+        memo: resp.get_memo().to_string(),
+    }))
+}
+
+#[derive(Deserialize)]
+struct JsonTransferRequest {
+    request_code: String,
+    amount: u64,
+}
+#[derive(Deserialize, Serialize)]
+struct JsonTransferResponse {
+    key_image: String,
+    tombstone: u64,
+}
+
+/// Performs a transfer from a monitor and subaddress. The target and amount are in the POST data.
+#[post(
+    "/monitors/<monitor_hex>/<subaddress_index>/transfer",
+    format = "json",
+    data = "<transfer>"
+)]
+fn transfer(
+    state: rocket::State<State>,
+    monitor_hex: String,
+    subaddress_index: u64,
+    transfer: Json<JsonTransferRequest>,
+) -> Result<Json<JsonTransferResponse>, String> {
+    let monitor_id =
+        hex::decode(monitor_hex).map_err(|err| format!("Failed to decode monitor hex: {}", err))?;
+
+    let mut req = mc_mobilecoind_api::ReadRequestCodeRequest::new();
+    req.set_b58_code(transfer.request_code.clone());
+    let resp = state
+        .mobilecoind_api_client
+        .read_request_code(&req)
+        .map_err(|err| format!("Failed reading request code: {}", err))?;
+    let public_address = resp.get_receiver();
+
+    let mut outlay = mc_mobilecoind_api::Outlay::new();
+    outlay.set_receiver(public_address.clone());
+    outlay.set_value(transfer.amount);
+
+    let mut req = mc_mobilecoind_api::SendPaymentRequest::new();
+    req.set_sender_monitor_id(monitor_id);
+    req.set_sender_subaddress(subaddress_index);
+    req.set_outlay_list(RepeatedField::from_vec(vec![outlay]));
+
+    let resp = state
+        .mobilecoind_api_client
+        .send_payment(&req)
+        .map_err(|err| format!("Failed to send payment: {}", err))?;
+
+    let receipt = resp.get_sender_tx_receipt();
+    Ok(Json(JsonTransferResponse {
+        key_image: hex::encode(receipt.get_key_image_list()[0].get_data()),
```

Comment:
This should return the whole list, something like `key_image: receipt.get_key_image_list().iter().map(|key_image| hex::encode(key_image.get_data()).collect()` should work if key_images is changed to `Vec<String>`.